Component::set_retry updates

e28eta opened the pull request:
# What does this implement/fix?

I was using `Component::set_retry` during some WIP development, and was both misled by the method documentation and discovered a minor bug in the timing of the first retry when the first execution of the function needs to be retried.

- clarification that `f` isn't executed immediately, it's called on the very next loop of the scheduler.
  - bug fix: #3225 caused `f` to be executed after `initial_wait_time`, it's back to doing it on the next scheduler loop
- `backoff_increase_factor` default is 1.0, *not* doubling the time
- clarification that `backoff_increase_factor` multiplies
- Behavior change / bug fix: do not apply `backoff_increase_factor` to the very first `initial_wait_time`
- Behavior change / bug fix: protect against negative or zero `backoff_increase_factor`. I have no idea what a caller would expect with a `backoff_increase_factor <= 0`, but adding protection against that case before dividing by `backoff_increase_factor`. I honestly don't know why a client would want a `backoff_increase_factor` that's a fraction close to zero (the time between retries gets significantly shorter each time), so arbitrarily picking 0.0001 as cutoff for disallowed values.


This function is not being used in esphome/esphome, but it's entirely possible that an external component uses it.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other: **developer documentation update**

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** n/a

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

e28eta committed:
Reword documentation on set_retry to match observed behavior

Looking at the PR, it looks like some aspects of this method documentation that confused
me were possible behavior changes between the time the PR / docs were first written and
what actually landed in the dev branch.

- `f` is not called immediately
- `backoff_increase_factor` default is *not* doubling the time
- s/increased/multiplied/ on the backoff_increase_factor to disambiguate between
additive/multiplicative increase

e28eta commented:
Unless I've made a mistake, I feel very confident you'll want this change. 

I have a follow up change I'd like to propose, which I'm less sure of and so I've split them into separate pieces of work: I think the retry function `f` should receive the number of attempts remaining as a parameter. Since I think this PR  is a good way to write the docs for this method I've used this as the base of my change. 

I'll wait until this PR lands to open the next PR in esphome, but if you'd like to take a look at it right now while you've got context on the `set_retry` method it's at e28eta/esphome#1.

e28eta commented:
I *did* make a mistake. `f` **is** called "immediately" (next loop): it's just that my wait time was less than the time it took for the retry handler to get called.

e28eta commented:
I *did* make a mistake. `f` **is** called "immediately" (next loop): it's just that my wait time was less than the time it took for the retry handler to get called.

e28eta committed:
Fix documentation: f is called immediately; fix code: first retry after `initial_wait_time`

My "rewrite" of the docs that claimed this doesn't execute `f` immediately was wrong, it
**is** being executed on the next scheduler loop, I was just trying to use an interval
that's less than the time for the initial setup tick to finish and didn't realize it.

This is still different from original docs in that the backoff_increase_factor default
was incorrectly documented, and I think specifying the increase happens via multiplication
is useful.

This commit *also* changes the code to match the originally documented backoff behavior.
As previously implemented, the 1st delay was (initial_wait_time * backoff_increase_factor),
contrary to the documentation. I think the as-documented behavior matches the method
signature the best, and I think this is a reasonably implemented bugfix.

e28eta committed:
Added: note about overflow, and fixed parameter docs

* With enough retries and a large enough backoff factor, it'd be possible to overflow the
`interval` between retries. Added note in documentation about this, and I think I got
the formula correct. If I did my math right, UINT32_MAX milliseconds is ~71 minutes, and
that actually seems like a plausible number to reach.
* As I've written this documentation, the first, immediate execution is *not* a "retry"
of `f`. The first "retry" happens after `initial_wait_time`, so I'm changing the doc
for `max_attempts`.
* similar for `backoff_increase_factor`: it isn't used for the first, immediate execution,
nor the first "retry" after `initial_wait_time`. It's used for the second (and later)
retries

e28eta committed:
Fix formatting error with long line, thanks clang-format

e28eta commented on esphome/core/scheduler.cpp:
```diff
@@ -84,14 +84,23 @@ void HOT Scheduler::set_retry(Component *component, const std::string &name, uin
-  item->interval = initial_wait_time;
+  // first execution happens on next loop, adjust `interval` so that 2nd execution happens
+  // after initial_wait_time (give or take some integer truncation)
+  item->interval = initial_wait_time / backoff_increase_factor;
```
Per brief discussion in #devs discord channel (thanks jesserockz!), I'd discovered that `set_retry("reset", 1000, 3, f, 5.0)` executed at `(now, +5000, +25000)`, because the `initial_wait_time` interval for the retry is being multiplied by the `backoff_increase_factor` when rescheduling. My expectation based on original documentation and the method signature would be `(now, +1000, +5000)`

This is the least invasive way I could think to maintain the documented behavior and fix the implementation.

Potential downsides:
* integer truncation means the intervals may not match what the dev expects, but we've already warned about the scheduler being imprecise...
* the log message for the initial execution shows `Running retry 'reset' with interval=200 last_execution=4294966376 (now=103)`. I found the `interval=<initial_wait_time>` in that log message misleading *in general* for the 1st execution of `f` since it's actually executing on the next scheduler loop, and showing `interval=<initial_wait_time / backoff_increase_factor>` may be even more misleading, but it's only when Very Verbose logging is turned on and it fixes the behavior... 🤷‍♂️

e28eta replied:
Good news, this divide and then multiply is no longer needed - and that's great because when integer truncation turned `initial_wait_time / backoff_increase_factor` into `0`, this was broken.

e28eta commented:
I still believe this is worth merging. Initially it had some problems, but afaik the code and PR description accurately describe the issues that this addresses

> On Jun 19, 2022, at 5:46 PM, github-actions[bot] ***@***.***> wrote:
> 
> ﻿
> There hasn't been any activity on this pull request recently. This pull request has been automatically marked as stale because of that and will be closed if no further activity occurs within 7 days. Thank you for your contributions.
> 
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.

jesserockz commented:
There are conflicts on this PR as the retry schedule code has been changed a little since you opened.

e28eta commented:
> There are conflicts on this PR as the retry schedule code has been changed a little since you opened.

Oh drat. That’s what I get for replying to the github email instead of opening the PR and using my eyes. 🤦‍♂️  Thanks for letting me know, I’ll take a look later this week!

e28eta committed:
Fix scheduler to execute `set_retry` function immediately, as documented

PR #3225 reworked `set_retry`, and made it do what I'd originally thought it did: execute
the function after `initial_wait_time`. However, the documentation and previous behavior
was to execute the function immediately, and then start the delay process.

Good news: the restructuring from #3225 made it easier to only apply the backoff factor
on the third & later executions.

e28eta commented on esphome/core/scheduler.cpp:
```diff
@@ -87,8 +87,10 @@ static void retry_handler(const std::shared_ptr<RetryArgs> &args) {
+  // second execution of `func` hapens after `initial_wait_time`
   args->scheduler->set_timeout(args->component, args->name, args->current_interval, [args]() { retry_handler(args); });
+  // backoff_increase_factor applied to third & later executions
+  args->current_interval *= args->backoff_increase_factor;
```
re-ordering the `set_timeout` and `current_interval = ` fixes the long-standing bug where `backoff_increase_factor` is applied too soon.

with the fix: `set_retry("reset", 1000, 3, f, 5.0)` executes at `(now, +1000, +5000)` instead of `(now, +5000, +25000)`

e28eta commented on esphome/core/scheduler.cpp:
```diff
@@ -111,7 +120,8 @@ void HOT Scheduler::set_retry(Component *component, const std::string &name, uin
 
-  this->set_timeout(component, args->name, initial_wait_time, [args]() { retry_handler(args); });
+  // First exectuion of `func` immediately
+  this->set_timeout(component, args->name, 0, [args]() { retry_handler(args); });
```
#3225 introduced a bug where the first execution of `func` happens after `initial_wait_time` instead of immediately. I'm changing the timeout value to 0 to fix that.

e28eta reviewed (commented):
Instead of testing in #3307 (like I'd done in the past), here's a basic gpio sensor lambda that I used to make sure it's compiling and scheduling correctly:

```yaml
binary_sensor:
  - platform: gpio
    id: gpio_gpiobinarysensor
    pin: 
      number: A0
      mode:
        input: true
        pullup: true
      inverted: true
    name: "A0 GPIO"
    filters:
      - delayed_on: 100ms
      - delayed_off: 100ms
    on_press:
      then:
        - lambda: |-
            ESP_LOGE("some-tag", "gpio pressed");
            App.scheduler.set_retry(id(gpio_gpiobinarysensor), "a0-retrying", 100, 6, []() {
              ESP_LOGE("some-tag", "custom lambda running");
              return RetryResult::RETRY;
            }, 5.0f);
```

and the log output, showing executions at `(now, +100, +500, +2500, +12500, +62500)`, which is what I'd expect for `initial_wait_time = 100`, `max_attempts = 6`, and `backoff_increase_factor = 5.0`

```
[18:21:17][D][binary_sensor:036]: 'A0 GPIO': Sending state ON
[18:21:17][E][some-tag:043]: gpio pressed
[18:21:17][VV][scheduler:105]: set_retry(name='a0-retrying', initial_wait_time=100, max_attempts=6, backoff_factor=5.0)
[18:21:17][VV][scheduler:026]: set_timeout(name='retry$a0-retrying', timeout=0)
[18:21:17][VV][scheduler:206]: Running timeout 'retry$a0-retrying' with interval=0 last_execution=11948 (now=11972)
[18:21:17][E][some-tag:045]: custom lambda running
[18:21:17][VV][scheduler:026]: set_timeout(name='retry$a0-retrying', timeout=100)
[18:21:17][VV][scheduler:206]: Running timeout 'retry$a0-retrying' with interval=100 last_execution=11976 (now=12076)
[18:21:17][E][some-tag:045]: custom lambda running
[18:21:17][VV][scheduler:026]: set_timeout(name='retry$a0-retrying', timeout=500)
[18:21:18][VV][scheduler:026]: set_timeout(name='OFF', timeout=100)
[18:21:18][VV][scheduler:206]: Running timeout 'OFF' with interval=100 last_execution=12424 (now=12529)
[18:21:18][D][binary_sensor:036]: 'A0 GPIO': Sending state OFF
[18:21:18][VV][scheduler:206]: Running timeout 'retry$a0-retrying' with interval=500 last_execution=12080 (now=12581)
[18:21:18][E][some-tag:045]: custom lambda running
[18:21:18][VV][scheduler:026]: set_timeout(name='retry$a0-retrying', timeout=2500)
[18:21:20][VV][scheduler:206]: Running timeout 'retry$a0-retrying' with interval=2500 last_execution=12585 (now=15085)
[18:21:20][E][some-tag:045]: custom lambda running
[18:21:20][VV][scheduler:026]: set_timeout(name='retry$a0-retrying', timeout=12500)
[18:21:33][VV][scheduler:206]: Running timeout 'retry$a0-retrying' with interval=12500 last_execution=15089 (now=27589)
[18:21:33][E][some-tag:045]: custom lambda running
[18:21:33][VV][scheduler:026]: set_timeout(name='retry$a0-retrying', timeout=62500)
[18:21:46][VV][scheduler:206]: Running interval '' with interval=60000 last_execution=4294948343 (now=41052)
[18:22:35][VV][scheduler:206]: Running timeout 'retry$a0-retrying' with interval=62500 last_execution=27593 (now=90093)
[18:22:35][E][some-tag:045]: custom lambda running
```

e28eta committed:
Merge branch 'dev' into set-retry-docs

e28eta committed:
Merge branch 'dev' into set-retry-docs

jesserockz committed:
Merge branch 'dev' into set-retry-docs
